Adjust hash reannounce based on queue size , test for DHTHashAnnouncer

kaykurokawa opened the pull request:
This patch fixes https://github.com/lbryio/lbry/issues/432

It was discovered that after running a reflector node with 16000 hashes, the hash queue (hashes waiting to be announced) was growing at a very fast rate. After 18 hours of running, the hash queue had about 160,000 hashes (see below debug lines from reflector4). 

2017-02-08 00:04:00,459 DEBUG    lbrynet.core.server.DHTHashAnnouncer:67: There are now 14 hashes remaining to be announced
2017-02-08 18:41:07,384 DEBUG    lbrynet.core.server.DHTHashAnnouncer:67: There are now 164043 hashes remaining to be announced

Since each hash takes about half a second to announce, this means it would take 80,000 seconds (or about 22 hours) for any new hash that got put into the queue to be announced , and the condition would obviously worsen after running the node for longer. So basically the large hash queue size was preventing newly published blobs from being re announced until many days later, depending on how long the node had been running.

This patch makes the hash reannounce time adjustable based on the queue size. The default reannounce time is 1 hour, but if it detects that it would take more than 1 hour to announce the entire hash queue, it sets the reannounce time to a conservative estimate of how long it would take to clear the queue ( it assumes it takes 1 second to announce a single hash ). Note that if a node contains 16000 hashes, the reannounce time will be on average 16000 seconds or about 4.4 hours.

We also add a very basic unit test for DHTHashAnnouncer, that should be expanded. 

Some variables in DHTHashAnnouncer and DHTHashSupplier that should be constants, have been turned into constants. 

I'm currently running this on reflector4 to test this patch.

jobevers commented on lbrynet/core/server/DHTHashAnnouncer.py:
```diff
@@ -10,6 +10,10 @@
 
 
 class DHTHashAnnouncer(object):
+    callLater = reactor.callLater
```
can you do the utils.callLater change that we talked about earlier?

jobevers commented on lbrynet/core/server/DHTHashAnnouncer.py:
```diff
@@ -87,12 +90,32 @@ def announce():
     def hashes_to_announce(self):
         pass
+
+    """
```
put this comment in the docstring under the function

jobevers commented on tests/unit/core/server/test_DHTHashAnnouncer.py:
```diff
@@ -0,0 +1,50 @@
+from twisted.internet import defer,task
+import logging
+import sys
+import os,binascii
```
import should be ordered by

```
system

third party

local
```

and alphabetical within each grouping.  Don't have multiple imports on the same line.  Instead do
```
import binascii
import os
```

jobevers commented on lbrynet/core/server/DHTHashAnnouncer.py:
```diff
@@ -87,12 +90,32 @@ def announce():
+        queue_size = self.hash_announcer.hash_queue_size()+num_hashes_to_announce
+        reannounce = max(self.MIN_HASH_REANNOUNCE_TIME,
+                            queue_size*self.SINGLE_HASH_ANNOUNCE_DURATION)
+        return time.time() + reannounce
```
I'm not convinced that this is a totally effective solution as it largely moves the queue from being in the queue data structure to being in the database.  The main difference is that the database will only have on entry for each blob so you decrease the chances of having repeated items in the queue - which is an improvement.

A better long term solution would be something along the lines of https://github.com/lbryio/lbry/pull/435/files - which is to record when a blob was last announced (instead of when to next announce it) and have a data structure that doesn't allow for repeated blobs.  As that requires a more substantial change, I'm okay with putting this in place for now but we should follow up with a more complete fix.

kaykurokawa replied:
Yea this patch is pretty bad in that we are increasing the dependency between BlobManager and DHTHashAnnouncer. BlobManager should not really need to handle announce time related logic at all except maybe to read and write the database entry for it. I agree that your patch would be be the direction to go next.

jobevers commented:
Question from tech meeting - can we prepend newly reflected blobs to the the queue instead of appending them.

kaykurokawa commented:
I don't see why not, we are using a deque .

kaykurokawa commented:
Rebased to master 

Added immediate announce option, (when calling immediate_announce, the hash gets added to the front of the queue, instead of the back) 

Fixed doc string formatting, fixed unit test import formatting

Added changelog

jobevers commented on lbrynet/core/server/DHTHashAnnouncer.py:
```diff
@@ -56,7 +56,7 @@ def _announce_available_hashes(self):
         return dl
 
-    def _announce_hashes(self, hashes):
+    def _announce_hashes(self, hashes, immediate=False):
```
👍

kaykurokawa commented:
seems to be failing test tests.unit.lbryfilemanager.test_EncryptedFileCreator.CreateEncryptedFileTest, 

there is some weired dependency on blob manager that's causing it to fail I guess.

kaykurokawa commented:
Fixed the failing unit test tests.unit.lbryfilemanager.test_EncryptedFileCreator.CreateEncryptedFile

jobevers commented:
LGTM

kaykurokawa commented:
rebase to master
added a fix to functional test (mock Announcer class now requires function hash_queue_size()
